client/lib/menu-data/menu-data.js

```diff
@@ -411,7 +434,47 @@ MenuData.prototype.saveMenu = function( menu, callback ) {
 			this.change( { reset: true } );
 			this.emit( 'saved' );
 			callback && callback( null, parsedMenu );
-		} );
+		}
+	);
+}
+
+MenuData.prototype.saveMenu = function( menu, callback ) {
+	if ( ! menu ) {
+		menu = this.find( { id: this.lastChangedMenuID } );
+	}
+
+	if ( ! this.isValidMenu( menu ) ) {
+		callback && callback( new Error( 'Invalid menu' ) );
+		debug( 'saveMenu: fail' );
+		return;
+	}
+
+	menu = cloneDeep( menu );
+
+	if ( menu.id === this.getDefaultMenuId() ) {
+		return this.saveDefaultMenu();
+	}
+
+	this.restoreServerIDs( menu );
+
+	debug( 'saveMenu', menu );
+
+	const self = this;
+
+	/*
+	Because of how postEditStore works, we need to fire promises one after another.
+	 */
+	menu.items.filter(
+		isInjectedNewPageItem
+	).reduce(
+		( previousNewPagePromise, menuItem ) => previousNewPagePromise.then( () => self.createNewPagePromise( menuItem, self.siteID ) ),
+		Promise.resolve()
+	)
+	.catch( error => {
+		self.emit( 'error', i18n.translate( 'There was a problem saving your menu.' ) );
+		debug( 'Error', error );
+	} )
+	.then( this.sendMenuToApi.bind( this, menu, callback ) );
```

Comment:
Here's a suggestion on how to make the above more readable (or, IMO, more self-explaining):

```js
	/**
	 * Consumes a list by
	 * - creating a promise from an item
	 * - waiting until its resolution
	 * - proceeding to the next item
	 *
	 * Returns a promise that wait for the last item promise to be resolved.
	 */
	const makePromiseSequence = ( list, fn ) => list.reduce(
		( acc, curr ) => acc.then( fn.bind( null, curr ) ),
		Promise.resolve()
	);

	// Because of how postEditStore works, we need to create pages one after
	// another.
	const pendingPageItems = menu.items.filter( isInjectedNewPageItem );
	const createdPages = makePromiseSequence(
		pendingPageItems,
		item => this.createNewPagePromise( item, this.siteID )
	);

	createdPages.catch( error => {
		this.emit( 'error', i18n.translate( 'There was a problem saving your menu.' ) );
		debug( 'Error', error );
	} )
	.then( this.sendMenuToApi.bind( this, menu, callback ) );
```